File: LeetCode/Python/Minimum Array End.py
"""
  Minimum Array End
  URL: https://leetcode.com/problems/minimum-array-end
  Difficulty: Medium
  Topic : Bit Manipulation
"""

class Solution:
    def minEnd(self, n: int, x: int) -> int:
        xBinString = f"0{x:b}"[::-1]
        binaryPow = []
        for i in range(len(xBinString)):
            if xBinString[i] == "0":
                binaryPow.append(i)
        nBinString = f"{n-1:b}"[::-1]
        result = 0
        for i in range(len(nBinString)):
            if nBinString[i] == "1":
                nPow = binaryPow[i] if len(binaryPow) > i else (binaryPow[-1] + i - len(binaryPow) + 1)
                result += 2 ** nPow
        return result + x

File: LeetCode/Medium/Minimum Array End.swift
/*
  Minimum Array End
  URL: https://leetcode.com/problems/minimum-array-end
  Difficulty: Medium
  Topic : Bit Manipulation
*/

import Foundation

class Solution {
    func minEnd(_ n: Int, _ x: Int) -> Int {
        let xBinString = "0\(String(x, radix:2))".reversed()
        var binaryPow: [Int] = []
        for item in xBinString.enumerated() {
            if item.element == "0" {
                binaryPow.append(item.offset)
            }
        }
        let nBinString = String(n-1, radix:2).reversed()
        var result = 0
        for item in nBinString.enumerated() {
            if item.element == "1" {
                var nPow = 0
                if binaryPow.count > item.offset {
                    nPow = binaryPow[item.offset]
                } else {
                    nPow = binaryPow.last! + item.offset - binaryPow.count + 1
                }
                result += Int(pow(2.0,Double(nPow)))
            }
        }
        return x + result
    }
}
